akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala

```diff
@@ -303,6 +303,83 @@ object GraphStages {
     override def toString: String = s"SingleSource($elem)"
   }
 
+  final class FutureFlattenSource[T, M](
+    val future: Future[Graph[SourceShape[T], M]])
+    extends GraphStageWithMaterializedValue[SourceShape[T], Future[M]] {
+
+    ReactiveStreamsCompliance.requireNonNullElement(future)
+
+    val out = Outlet[T]("FutureFlattenSource.out")
+    override val shape = SourceShape(out)
+
+    override def initialAttributes = DefaultAttributes.futureSource
+
+    override def createLogicAndMaterializedValue(attr: Attributes): (GraphStageLogic, Future[M]) = {
+      val materialized = Promise[M]()
+
+      val logic = new GraphStageLogic(shape) with InHandler with OutHandler {
+        private val sinkIn = new SubSinkInlet[T]("FutureFlattenSource.in")
+
+        override def preStart(): Unit = {
+          val cb = getAsyncCallback[Try[Graph[SourceShape[T], M]]](onFutureSourceCompleted).invoke _
+          future.onComplete(cb)(ExecutionContexts.sameThreadExecutionContext)
+        }
+
+        // initial handler (until future completes)
+        setHandler(out, new OutHandler {
+          def onPull(): Unit = {}
+
+          override def onDownstreamFinish(): Unit = {
+            materialized.tryFailure(new RuntimeException("Downstream cancelled before future source completed"))
```

Comment:
Change it to "Downstream cancelled before future source completed. There is not available materialized value because the inner source was not yet materialized."

The reason is that people might expect to see the materialized value anyway, even though the downstream cancelled. We cannot provide it of course because we have not materialized it yet, but this might not be evident for users.

Alternatively, we can ensure that the inner source is always materialized:

```scala
// onDownstreamFinish
val matValueFuture = future.map(_.to(Sink.cancel).run())
materialized.completeWith(matValueFuture)
```

I am leaning towards this second option as whether you get an exception or not is racy.